.github/workflows/ci-master.yml

```diff
@@ -3,63 +3,184 @@ on:
+    name: Create Source Distribution
+    runs-on: ubuntu-18.04
+    env:
+      ARTIFACT_DIR: source
     steps:
     - name: Checkout
-      uses: actions/checkout@v1
-    - name: Install dependencies
+      uses: actions/checkout@v2
+    - name: Install Required Packages
       run: |
-        sudo apt update
-        sudo apt-get install -y ccache python3-zmq
-    - name: Build dependencies
-      run: make -C depends -j$(nproc) HOST=x86_64-unknown-linux-gnu
-    - name: Build Zcoin
+        sudo apt-get update
+        sudo apt-get install -y libboost-all-dev libdb-dev libdb++-dev libevent-dev
+    - name: Create Distribution Tarball
       run: |
         ./autogen.sh
-        ./configure --enable-exodus --enable-tests --with-comparison-tool=no --prefix=$(pwd)/depends/x86_64-unknown-linux-gnu
+        ./configure --with-incompatible-bdb
+        make dist
+    - name: Download Dependencies
+      run: make -C depends download
+    - name: Create Dependencies Tarball
+      run: tar -czf depends.tar.gz depends
+    - name: Prepare Files for Artifact
+      run: |
+        mkdir -p $ARTIFACT_DIR
+        mv depends.tar.gz zcoin-*.tar.gz $ARTIFACT_DIR
+    - name: Upload Artifact
+      uses: actions/upload-artifact@v1
+      with:
+        name: ${{ env.SOURCE_ARTIFACT }}
+        path: ${{ env.ARTIFACT_DIR }}
+  build-linux:
+    name: Build for Linux
+    needs: create-source-distribution
+    runs-on: ubuntu-18.04
+    env:
+      ARTIFACT_DIR: linux-binaries
+      TEST_LOG_ARTIFACT_DIR: test-logs
+    steps:
+    - name: Getting Source
+      uses: actions/download-artifact@v1
+      with:
+        name: ${{ env.SOURCE_ARTIFACT }}
+    - name: Extract Archives
+      run: |
+        tar -xzf depends.tar.gz
+        tar -xzf zcoin-*.tar.gz --strip-components=1
+      working-directory: ${{ env.SOURCE_ARTIFACT }}
+    - name: Install Required Packages
+      run: |
+        sudo apt-get update
+        sudo apt-get install -y python3-zmq
+    - name: Build Dependencies
+      run: make -C depends -j$(nproc)
+      working-directory: ${{ env.SOURCE_ARTIFACT }}
+    - name: Build Zcoin
+      run: |
+        ./configure --disable-jni --enable-exodus --enable-tests --with-comparison-tool=no --prefix=$(realpath depends/x86_64-pc-linux-gnu)
         make -j$(nproc)
-    - name: Unit Tests
+      working-directory: ${{ env.SOURCE_ARTIFACT }}
+    - name: Run Unit Tests
       run: make check
-    - name: RPC Tests
+      working-directory: ${{ env.SOURCE_ARTIFACT }}
+    - name: Run RPC Tests
       env:
         TIMEOUT: 600
       run: qa/pull-tester/rpc-tests.py -extended
+      working-directory: ${{ env.SOURCE_ARTIFACT }}
+    - name: Prepare Files for Artifact
+      run: |
+        mkdir -p $ARTIFACT_DIR
+        mv $SOURCE_ARTIFACT/src/{zcoin-cli,zcoin-tx,zcoind,qt/zcoin-qt} $ARTIFACT_DIR
+    - name: Upload Artifact
+      uses: actions/upload-artifact@v1
+      with:
+        name: linux-binaries
+        path: ${{ env.ARTIFACT_DIR }}
```

Comment:
I believe we should archive using tarball before upload to preserve permissions.

Reply:
Yes, I agreed. But it is also ugly that it needs to extract files twice. We should try to find another solution.

Reply:
After investigating, we don't have simple ways to do. And the main purpose of binaries is used for development then I think we should let it be like this for now and create an issue instead.

Reply:
https://github.com/zcoinofficial/zcoin/issues/833